「コメントを付けられるようにする」の課題

cellotak opened the pull request:
「[コメントを付けられるようにする \| FBC](https://bootcamp.fjord.jp/practices/169)」の課題として、books_appに日報投稿機能、およびコメント機能を追加します。

レビューお願い致します。

以下GitHubのみ見る方のために、要件の内容の説明と、実際の機能の説明(スクリーンショット付き)を記載します。

必須要件、歓迎要件、注意すべき点はそれぞれ以下で、すべて完了しています。

## 必須要件
<details>

- [x]  本（books）のCRUDを作ったときの復習として日報（reports）が投稿できるようにする（reportsのCRUDを作る。入力項目はタイトルと本文だけで良い。created_atの日付を日報の投稿日とする）
- [x] 日報を更新・削除できるのはその日報を投稿した本人のみ（本のCRUDは特に制限なし）
- [x] 本と日報の各詳細画面からコメントを投稿できるようにする。
- [x] コメントを削除できるのはそのコメントを投稿した本人のみ
- [x] 投稿したコメントは本と日報の各詳細画面から確認できる。投稿した内容に加えて、投稿したユーザーの名前（未入力ならメアド）と投稿日時も表示する
- [x] 本と日報の各詳細画面でコメントが投稿できている様子をスクショする
- [x] rubocopとerblintをパスさせる（要スクリーンショット）
- [x] bookに付くコメントとreportに付くコメントを共通のコードで動くようにする。

</details>

## 歓迎要件
<details>

- [x] コメントが1件もない場合はその旨を画面に表示する
- [x] HTML5の必須チェックが効かない場合を考慮して、バリデーションエラー時のレスポンスも返せるようにする
- [x] コメントの編集ができるようにする

</details>

## 注意すべき点
<details>

- [x] コメント入力フォームはHTML5の必須チェック機能を使うと、コメント未入力で投稿された時の考慮が不要となる
- [x] scaffoldの自動生成機能を使う場合は、JSONレスポンスのように使う予定のない機能は削除した状態でコミットする
- [x] 本や日報が削除されたらコメントはどうなるか？ユーザーが退会したらコメントはどうなるか？を検討する
- [x] 日報の投稿日やコメントの投稿日時はタイムゾーンの扱いに注意する（日本時間じゃないかも？）
- [x] i18nの機能を使って日本語表示する（英語化は考慮不要）

</details>

## 機能の説明

### 日報（reports）の投稿機能
日報のCURDを実装しました。

- 日報の一覧画面にすべてのユーザの日報が表示されます。
    <details>
    
    「http://127.0.0.1:3000/reports」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/dd5a01f0-1ae1-41a5-83f7-aa6055f6ce96)
    
    </details>


- 「この日報」を表示というリンクをクリックすると日報の詳細画面が表示されます。自分の日報には編集するためのリンクや削除ボタンが表示されており、編集と削除が行えます。

    <details>
    
    「http://127.0.0.1:3000/reports/1」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/8d26fea9-b32d-4e19-8d63-9229b909e210)
    
    </details>



- 他人の日報は閲覧できますが、編集のリンクや削除ボタンがないため、編集や削除はできません。また、リクエストを直接送っても編集や削除はできないようになっています。

    <details>
    
    「http://127.0.0.1:3000/reports/500」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/9371dea4-bc85-4a8f-b3a4-1c7a5ac60be6)
    
    </details>

- 日報一覧画面の一番下の「日報の新規作成」のリンクをクリックすると、新規作成画面が表示されます。入力項目はタイトルと本文だけです。
    <details>
    
    「http://127.0.0.1:3000/reports/new」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/a5e9a263-ed80-4593-895a-b8327c38c01f)
    
    </details>


- 内容を入力して登録ボタンを押すと、日報が作成され、日報詳細画面に移動します。投稿日時(日時は日本時間)、投稿したユーザの氏名、タイトル、内容が表示されます。
    <details>
    
    「http://127.0.0.1:3000/reports/501」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/3877bf7a-d9da-48a6-8ab8-85aac9e86f7e)
    
    </details>


- ユーザの詳細画面に、そのユーザが投稿した日報一覧が表示されるようになっています。
    <details>
    
    「http://127.0.0.1:3000/users/1」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/28e22b62-115c-4a1a-9774-57615354b321)
    
    </details>


### コメントの投稿機能

- 本の詳細画面にコメント投稿欄と、コメント一覧の欄が追加されています。
    <details>
    
    「http://127.0.0.1:3000/books/1」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/cb8f0f52-d6c3-455c-89a5-85fbdf19ed4a)
    
    </details>


- コメントの投稿フォームに文字を入力した状態で「コメントの投稿」ボタンをクリックすると、「コメントが作成されました。」というメッセージがでてきて、コメント一覧に作成したコメントが表示されます。
コメントの投稿日時 (日時は日本時間)、投稿者、内容が表示されます。
    <details>
    
    「http://127.0.0.1:3000/books/1」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/91ca7528-a766-40c6-9437-fdd6cdc254c8)
    
    </details>


- コメントを投稿したユーザが名無しの場合、「投稿者」の欄にメールアドレスが表示されるようになっています。
    <details>
    
    「http://127.0.0.1:3000/books/1」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/4f4a2e51-5da6-4997-8f83-f37070f3bd74)
    
    </details>


- コメント一覧のうち、自分が投稿したコメントには削除ボタンが表示されていて、その削除ボタンをクリックすることでコメントを削除できます。
自分以外が投稿したコメントには削除ボタンは表示されません
    <details>
    
    「http://127.0.0.1:3000/books/1」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/91ca7528-a766-40c6-9437-fdd6cdc254c8)
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/4dc22102-c4d2-486a-aeef-abe433554b01)
    
    </details>



- 現在投稿されているコメントがない場合、コメント一覧には「コメントがありません」と表示されています。
    <details>
    
    「http://127.0.0.1:3000/books/2」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/005b8342-849d-4cad-8b4f-7a85bd36293c)
    
    </details>


- 「このコメントの詳細表示」というリンクをクリックすると、コメントの詳細画面が表示されます。自分が投稿したコメントの場合、「このコメントを編集」というリンクとコメントの削除ボタンが追加で表示されます。
    <details>
    
    「http://127.0.0.1:3000/books/1/comments/1」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/60f7c10d-8e34-4ccd-b003-ef4fdd4f7dfc)
    
    </details>


- 「このコメントを編集」というリンクをクリックすると、コメントの編集画面が表示され、コメントを編集することができます。
削除ボタンはコメント一覧に表示されているものと同様、コメントが削除できます。
    <details>
    
    「http://127.0.0.1:3000/books/1/comments/1/edit」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/92623e51-a65d-45a8-87cf-08da2adc6e1f)
    
    </details>


- 自分が投稿していないコメントの詳細画面には「このコメントを編集」というリンクとコメントの削除ボタンは表示されていません。そのため編集や削除は基本的にはできませんが、直接リクエストを送っても編集や削除はできないようになっています。
    <details>
    
    「http://127.0.0.1:3000/books/1/comments/2」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/88292f3d-a47e-4bca-8cf2-128d18608717)
    
    </details>


- コメント入力欄が空のまま「コメントを投稿」ボタンをクリックすると、HTML5の必須チェック機能により、メッセージが表示されて投稿ができないようになっています。
    <details>
    
    「http://127.0.0.1:3000/books/1」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/f25ad823-2bd5-42fe-999b-82da39f4827e)
    
    この機能は `app/views/comments/_form.html.erb`において以下のようにtext_areaにrequired属性を付けることで実現しています。
    
    ```erb
      <div>
        <%= form.label :content, style: "display: block" %>
        <%= form.text_area :content, required: true %>
      </div>
    
    ```
    
    仮にHTML5の必須チェックが効かない場合を想定して、`required: true` の記述を削除して空のコメントを投稿してみると、エラーメッセージが表示され、コメントは投稿されません。
    
    「http://127.0.0.1:3000/books/1」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/56cfa202-5800-41e3-b142-b466d42901f6)
    
    </details>


- 日報の詳細画面に対しても同様にコメント投稿欄とコメント一覧が追加されています。機能は本に対するコメント機能と同様のため説明は省略します。
    <details>
    
    「http://127.0.0.1:3000/reports/1」
    
    ![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/2f4ec3ac-44d2-44e9-a0a4-c41ad67268ce)
    
    </details>


## rubocopとerblintのチェック
rubocopとerblintはともにパスしています。

<details>

![image](https://github.com/cellotak/fjord-books_app-2023/assets/65953037/26aeee80-13c1-47c2-ba57-1545fd782e01)

</details>

## プラクティスページにリンクされている情報以外で参考にした情報
<details>

- [Rails scaffoldについて](https://zenn.dev/goldsaya/articles/02360f3fcb89e2)
- [TIL/99\_etc/20200902\_polymorphoric\_explanation\.md at master · miketa\-webprgr/TIL](https://github.com/miketa-webprgr/TIL/blob/master/99_etc/20200902_polymorphoric_explanation.md)
- [【Rails】ポリモーフィック関連のコントローラについて \| Mt\_LIFE Blog](https://show-time-blog.com/it-knowledge/ruby/1141/#toc10)

</details>

cellotak committed:
reportモデルをscaffoldで作成、およびuserモデルとの関連付け

cellotak committed:
user_idがcurrent_user.idになるように変更

cellotak committed:
user_idをcurrent_user.idにする実装方法を変更

cellotak committed:
ビューを編集中

cellotak committed:
日報に関するビューを整え、i18n化

cellotak committed:
currentユーザに紐づいたreportのみedit,update,destroyができるように変更

cellotak committed:
ユーザの詳細画面にそのユーザの日報一覧が表示されるように変更

cellotak committed:
不必要なファイルを削除

cellotak committed:
commentモデル作成、ポリモーフィック関連付けをする

cellotak committed:
余計なコードを削除

cellotak committed:
commentモデルのルーティングを設定

morimorihoge commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,76 @@
+  private
+    # Use callbacks to share common setup or constraints between actions.
+    def set_report
+      @report = Report.find(params[:id])
```
こういうログインユーザーが紐付いているModelから検索する場合には、常にcurrent_userを起点に検索するのが良いです。  
そうすることで、予期せず他のユーザーのデータを検索してしまう可能性を回避できます。

例

```ruby
@report = current_user.reports.find(params[:id])
```

morimorihoge replied:
showは他のユーザーのモノが見たいということなので、これは不要でした :bow:

morimorihoge commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,76 @@
+
+    def is_current_user_report?
+      if current_user != @report.user
+        redirect_to reports_path
```
想定するケースとしては、エラーケースだと思うので、redirect_toのオプション引数でエラーメッセージとかを渡すようにするとユーザーに「なぜいきなりreports#indexにリダイレクトされたか」が表示できて便利かも

cellotak replied:
下記のようにしてエラーメッセージを表示するようにしました。

```ruby
redirect_to reports_path, alert: t('controllers.common.alert_unauthorized_operation')
```

morimorihoge commented on app/controllers/users_controller.rb:
```diff
@@ -7,5 +7,6 @@ def index
 
   def show
     @user = User.find(params[:id])
+    @reports = @user.reports
```
`#order`で新しい順とかに指定しなくても良い？`#limit`も

konaga-k replied:
https://github.com/cellotak/fjord-books_app-2023/blob/f952c9564a0978041de5bae3949792d83dcfe722/Gemfile#L86

kaminariが入っているので、件数を絞るのはkaminariのやり方でよいかなと
並び順を決めたほうがいいというのは、たしかにそう

cellotak replied:
kaminariで新しい順に表示するようにしました。

morimorihoge commented on config/routes.rb:
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  resources :reports
```
細かいですが、`routes.rb`は割と良く見るファイルになるので、記述を追加していくときにある程度まとまった整理された順番に記述されていると後々読みやすいです。

例えば、

* mountable engineの定義
* rootの定義
* devise等のGem系のDSL routing
* 当該アプリ用のrouting

みたいな。

ただ、記述する順序に意味がある場合もあるので、そういった場合には多少前後して書くこともあります（routingの定義が被った場合は後勝ちだったはず）

cellotak replied:
元々が
- mountable engineの定義
- devise等のGem系のDSL routing
- rootの定義
- 当該アプリ用のrouting
という順序だったため、それに則りました。

morimorihoge commented on db/migrate/20230914024438_create_reports.rb:
```diff
@@ -0,0 +1,11 @@
+    create_table :reports do |t|
+      t.references :user, null: false, foreign_key: true
+      t.string :title
+      t.text :content
```
null: falseは常につけられるかどうか検討すべきです。

イメージとしてはNULLを明示的に当該カラムにセットして使いたい場合を除きNULLは禁止してdefaultをつけるべき。

cellotak replied:
`:title`も`:content`も 「`NULL'を明示的にカラムにセットして使いたい」というわけではないので `null: false`としました。

konaga-k commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,76 @@
+
+    # Only allow a list of trusted parameters through.
+    def report_params
+      params.require(:report).permit(:user_id, :title, :content)
```
[MUST]
`:user_id` は[このあたり](https://github.com/cellotak/fjord-books_app-2023/pull/7#discussion_r1325650135)で話した通りpermitしちゃだめなので対象から外してほしいです

cellotak replied:
完全に忘れていました、、
対象から外しました。

konaga-k commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,76 @@
+    respond_to do |format|
+      if @report.save
+        format.html { redirect_to report_url(@report), notice: t('controllers.common.notice_create', name: Report.model_name.human) }
+        format.json { render :show, status: :created, location: @report }
```
[SHOULD]
`respond_to` や `format.json` はscaffoldで自動生成されたコードだと思います
もしjsonのレスポンスを返す必要がなければ、json対応のコードは不要になるので消してみてください

`respond_to` は要求された拡張子の種類に応じて処理を分けるためのメソッドですね
htmlだけ返したい場合は、特に使う必要はないやつです
https://api.rubyonrails.org/classes/ActionController/MimeResponds.html#method-i-respond_to

cellotak replied:
jsonのレスポンスを返す必要はないので、`respond_to`の記述を削除しました。

konaga-k commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,76 @@
+class ReportsController < ApplicationController
+  before_action :set_report, only: %i[ show edit update destroy ]
```
[MAY]
これ自体scaffoldが吐き出しているコードなんだと思いますが、 `%i[]` の括弧の中の余計な半角スペースはない方がよいかなと（`show`の直前、`destroy`の直後）
これ以外にも冗長な空行が挟まっている箇所などあり、こういうのは `rubocop -a` で修正される気がするので試してもらえると

cellotak replied:
rubocop通し忘れていました。
修正しました。

konaga-k commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,76 @@
+  before_action :is_current_user_report?, only: %i[ edit update destroy ]
+  # GET /reports or /reports.json
+  def index
+    @reports = Report.order(:id).page(params[:page])
```
[Q]
日報一覧画面での日報の並び順がidの昇順ですが、これは意図通りということであってますか？
事実上、作成日が古いものから先頭に並ぶと思います

ちゃんと意図があってやっている場合、あんまり見ない並び順と思ったので、軽く説明もらえると助かります
自動生成された部分で未検討だったなどであれば、一度確認お願いします

cellotak replied:
book一覧やuser一覧ではid順にしていたので、特に考えずにそのままid昇順にしてしまっていました。
日報の場合、作成日時が新しいものが先頭に表示されるべきだと思いますので、そのように変更しました。

konaga-k commented on app/views/reports/_report.json.jbuilder:
```diff
@@ -0,0 +1,2 @@
+json.extract! report, :id, :user_id, :title, :content, :created_at, :updated_at
```
[SHOULD]
https://github.com/cellotak/fjord-books_app-2023/pull/9#discussion_r1330849571 の関連の話ですが、
`.json.builder` についても、意図して生成したのでなければファイルごと消してもらえると

cellotak replied:
余計なファイルを削除しました。

konaga-k commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,76 @@
+      params.require(:report).permit(:user_id, :title, :content)
+    end
+
+    def is_current_user_report?
```
[SHOULD]
メソッド名が処理内容の誤解を招きそうです

`?` で終わるメソッドは、慣用的にtrueかfalseを返すことが期待されています
https://docs.ruby-lang.org/ja/latest/doc/symref.html#q
> def xx?
> この場合の「?」はメソッド名の一部分です。慣用的に、真偽値を返すタイプのメソッドを示すために使われます。

このメソッドの実際の処理内容とは意味が違ってくるので、メソッド名を考えてみてください〜
思いつかない場合は相談してもらえれば

konaga-k replied:
[MAY]
`def report_params` と `def is_current_user_report?` の順番を入れ替えると可読性が上がりそうです

`before_action` が連続して書かれているので、そこで呼ばれるメソッドの定義も連続していたほうがいいかなあと
https://github.com/cellotak/fjord-books_app-2023/blob/f952c9564a0978041de5bae3949792d83dcfe722/app/controllers/reports_controller.rb#L2-L3

cellotak replied:
ご指摘ありがとうございます。当初は真偽値を返すことを想定していたけど、書いているうちにそうではなくなってしまったのだと思われます。(自分でもなぜそうなったか思い出せないです)

一応 `ensure_current_user`というように変更してみました。

並び順についても対応しました。

konaga-k reviewed (commented):
動作確認してないです
眺めて気になったところだけコメントしました！
scaffoldが生成したコードをそのままにしてる部分があると思うので、いくつかコメント入れてます

cellotak committed:
エラーメッセージの追加及び軽微な修正

current_user以外の日報に対してedit, update,
destroyをしようとしたときにエラーメッセージを出すように変更

rubocopの修正を何点か反映

paramsとしてuser_idを許可しないように変更

cellotak committed:
route.rbの記述順を変更

cellotak committed:
reportのtitleとcontentにNOT NULL制約を追加

cellotak committed:
reports_controllerで自動生成されていたrespond_toを削除

cellotak committed:
reportの一覧のorderを日付が新しい順になるように変更

cellotak committed:
destroyアクションのrespond_toを削除

cellotak committed:
reports_controllerのメソッド名の変更、記述位置の変更

cellotak committed:
scaffoldで生成された余計なファイルと余計な記述を削除

cellotak committed:
翻訳ファイルの表記ミスの修正

cellotak committed:
rubocopの指摘事項修正

cellotak committed:
レポートのダミーデータを作成

cellotak committed:
Merge pull request #10 from cellotak/my-comments-add-comments-model

my-commentsブランチレビュー中の作業ブランチ

cellotak committed:
routes.rbのミスを修正

cellotak committed:
rails gでcomments_controllerとcommntsのviewを生成

cellotak committed:
comment_controllerを作成、createのみ

commentableがreportかbookかで切り替えられるようにモジュール作成

cellotak committed:
余計なテストファイルを削除

cellotak committed:
reportに対するコメント機能を追加、showアクションとビューを作成

reportのshow画面にも追加、commentのshow画面へのリンクが正しくない

cellotak committed:
リンクが正しく生成されない問題を解決

cellotak committed:
日報に対するコメント投稿機能を追加

投稿フォームは日報詳細画面に追加

cellotak committed:
日報とコメントの投稿日時を表示する機能追加

日本時間が表示されるようにタイムゾーンを日本に変更

cellotak committed:
コメント削除機能を追加

ただし、どのユーザのコメントでも消すことができる状態

cellotak committed:
コメント削除は投稿した本人のみができるように変更

cellotak committed:
コメント投稿者の名前が未登録の場合、メールアドレスを表示するように変更

cellotak committed:
日報とコメントの投稿日時の表示周りを変更

cellotak committed:
からのコメントを入力できないように設定、一部18n化

cellotak committed:
名前空間を使ってレポートに対するコメントのcontrollerを分割

cellotak committed:
本に関してもコメント機能を追加

cellotak committed:
コメントフォームのパーシャルを共通化

cellotak committed:
投稿フォームのi18n化

cellotak committed:
詳細画面の削除ボタンを条件分岐で切り替えるように変更

cellotak committed:
条件分岐文のバグを解消

cellotak committed:
userとcommentsに対するdependent: :destroyを設定

cellotak committed:
rubocopとerblintの指摘部分修正

cellotak committed:
コメント詳細画面の削除ボタンがコメント投稿者のみに表示されるように変更

cellotak committed:
コメント詳細画面のバグ修正

cellotak committed:
本に対するコメントをパーシャルビュー化

cellotak committed:
削除ボタンの条件分岐をヘルパーメソッドに定義して共通化

cellotak committed:
コメント一覧のパーシャルビューを日報詳細でも利用できるように変更

cellotak committed:
コメント機能のi18n化追加

cellotak committed:
rubocopとerblintの修正

cellotak committed:
validationでも空のコメントが投稿できないように設定

cellotak committed:
commentのパーシャルビューのファイル名を変更

cellotak committed:
commentのパーシャルビューのファイル名を変更

konaga-k commented on app/helpers/comments_helper.rb:
```diff
@@ -0,0 +1,12 @@
+    when 'Book'
+      book_comment_path(comment.commentable, comment)
+    end
+  end
```
ここ悩んでたと思うんですが、 `polymorphic_path` が使えそうですね

https://bootcamp.fjord.jp/practices/169#-2
https://api.rubyonrails.org/classes/ActionDispatch/Routing/PolymorphicRoutes.html

もし改善したい場合は見てもらうといいかなと

cellotak replied:
ちゃんとヒントが書いてあったんですね、、
ありがとうございます！
それを使おうと思います。

konaga-k commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,38 @@
+  def show; end
+
+  def create
+    @comment = @commentable.comments.build(comment_params)
```
nested routingをやめられないか口頭で案だけ出してみたんですが、ここが一番のネックになるのでnested routingでもいいのかなと宗旨変えしました

コメントのCRUDを浅いroutingにするのも無理ではないです
もしやる場合は（※やってほしいという意味ではない）、bookへのコメントかreportへのコメントか区別できるように、 `#create` で `:commentable_type` , `:commentable_id` をパラメータ送信することになると思います

cellotak replied:
ありがとうございます。
今回はこのままnested routingにしますが、今後の参考にします。

konaga-k reviewed (commented):
通話した部分だけコメントしておきます

cellotak committed:
コメントの編集機能を追加

cellotak committed:
rubocopの修正

cellotak committed:
URL生成を自作helperからpolymorphic_pathに変更

cellotak committed:
コメント編集画面のタイトルを修正

cellotak committed:
他人のコメント詳細ページに「戻る」リンクを追加

cellotak committed:
rubocopの修正

morimorihoge commented on app/helpers/comments_helper.rb:
```diff
@@ -0,0 +1,12 @@
+    when 'Book'
+      Book.model_name.human.downcase
+    end
+  end
```
流石にこの程度なら `comment.commentable_type.constantize.model_name.human.downcase` とかでもいいかな？とは思いました。 `comment.commentable_type` にユーザー入力とかが入ってくると任意コード実行の脆弱性になりかねないのでアレですが、DBから取ってくるのが保証されてるなら信頼してもよさそう

konaga-k replied:
`comment.commentable.model_name.human.downcase` でもよさそう？

cellotak replied:
`comment.commentable.model_name.human.downcase` でできました！

yoshitsugu commented on app/views/reports/_report.html.erb:
```diff
@@ -0,0 +1,24 @@
+<div id="<%= dom_id report %>">
+    <p>
+    <strong><%= Report.human_attribute_name(:created_at) %>:</strong>
+    <%= report.created_at.strftime('%Y/%m/%d %H:%M:%S') %>
```
Railsにおいては `strftime` より `I18n.l` などが使われることが多いですね。
https://qiita.com/jnchito/items/831654253fb8a958ec25

cellotak replied:
`I18n.l` を使うように変更しました。

yoshitsugu reviewed (commented):
細かいところ一点だけコメントしました！

cellotak committed:
日時の表示をstrftimeからI18n.lに変更

cellotak committed:
commentableのモデル名取得のための独自helperを削除

yoshitsugu reviewed (approved):
OKです！
